### PR TITLE
Add transformations for infinite and semi infinite integral.

### PR DESCRIPTION
**Pull request by ashutosh-b-b**

### CODE REVIEW

**Comment by ashutosh-b-b** [comment]
@ChrisRackauckas  I applied this to Quadrature Algorithms. Can you check this further so that I can apply this to Cubature as well?

**Comment by ChrisRackauckas** [comment]
What do you mean? Cubature is just multi-dimensional quadrature.

**Comment by ashutosh-b-b** [comment]
No, I meant have I done correctly ? I added the check only to `AbstractQuadratureAlgorithm` part

**Comment by ChrisRackauckas** [comment]
It looks like it's written always assuming that lb and ub is scalar or length 1 which isn't correct, but when that's fixed it should be good to go.

**Comment by ashutosh-b-b** [comment]
> It looks like it's written always assuming that lb and ub is scalar or length 1 which isn't correct, but when that's fixed it should be good to go.

No, I have checked this works for vector inputs as well. Should I add a test case as well?

**Comment by ChrisRackauckas** [comment]
>No, I have checked this works for vector inputs as well. Should I add a test case as well?

Yes, on scalar and vector.

**Comment by ashutosh-b-b** [comment]
@ChrisRackauckas

File: src/Quadrature.jl [hunk 1]
```diff
@@ -43,6 +43,58 @@ function scale_x(ub,lb,x)
     (ub .- lb) .* x .+ lb
 end
 
+function transform_inf(t , p , f)
+  v(t) = t ./ (1 .- t.^2)
+  if t isa Number
+      j = ForwardDiff.derivative(v, t)
```
Comment (ChrisRackauckas) [inline src/Quadrature.jl:49]: can we just write down the analytical solution to the derivatives here?

File: src/Quadrature.jl [hunk 2]
```diff
@@ -43,6 +43,58 @@ function scale_x(ub,lb,x)
...
+  else
+      j = det(ForwardDiff.jacobian(x ->v(x), t))
+  end
+  f(v(t) , p)*(j)
+end
+
+function transformation_if_inf(prob)
+    if (!(prob.lb isa Number) && all(prob.lb .== -Inf) && all(prob.ub .== Inf)) || (prob.lb == -Inf && prob.ub == Inf)
```
Comment (ChrisRackauckas) [inline src/Quadrature.jl:67]: I wonder if this could be selectively applied to the directions which are Inf.

Comment (ashutosh-b-b) [reply src/Quadrature.jl:67]: Oh, does the transformation hold for that case?

Comment (ChrisRackauckas) [reply src/Quadrature.jl:67]: You can cache the indexes to transform and then if transform, then X otherwise identity.

**Comment by ashutosh-b-b** [comment]
@ChrisRackauckas

File: src/Quadrature.jl [hunk 3]
```diff
@@ -43,6 +43,85 @@ function scale_x(ub,lb,x)
...
+			j = (1 .+ t.^2 )/(1 .- t.^2).^2
+			return f(v_inf(t) , p)*(j)
+		elseif lb != -Inf && ub == Inf
+			a = lb
+			j = 1 ./ ((1 .- t).^2)
+			return f(v_semiinf(t , a) , p)*(j)
+		elseif lb == -Inf && ub != Inf
+			error("Semi Infinte integrals of limits 0 to Inf are supported")
+		end
+	end
```
Comment (ChrisRackauckas) [inline src/Quadrature.jl:68]: this should be a separate dispatch

File: src/Quadrature.jl [hunk 4]
```diff
@@ -43,6 +43,85 @@ function scale_x(ub,lb,x)
...
+	if lb isa Number && ub isa Number
+		if lb == -Inf && ub == Inf
+			j = (1 .+ t.^2 )/(1 .- t.^2).^2
+			return f(v_inf(t) , p)*(j)
+		elseif lb != -Inf && ub == Inf
+			a = lb
+			j = 1 ./ ((1 .- t).^2)
+			return f(v_semiinf(t , a) , p)*(j)
+		elseif lb == -Inf && ub != Inf
+			error("Semi Infinte integrals of limits 0 to Inf are supported")
```
Comment (ChrisRackauckas) [inline src/Quadrature.jl:66]: can't you just flip the integral to handle this case?

**Pull request merged**